src/server/hset_family.cc

```diff
@@ -166,8 +166,43 @@ OpStatus IncrementValue(optional<string_view> prev_val, IncrByParam* param) {
   param->emplace<int64_t>(new_val);
 
   return OpStatus::OK;
+}
+
+struct HSetCleanupCtx {
+  StringMap* sm;
+  DbSlice& db_slice;
+  const OpArgs& op_args;
+  std::string_view key;
+
+  void DelKeyIfEmpty() const;
 };
 
+void HSetCleanupCtx::DelKeyIfEmpty() const {
+  if (!sm->Empty()) {
+    return;
+  }
+
+  if (auto it = db_slice.FindMutable(op_args.db_cntx, key, OBJ_HASH); it) {
+    it->post_updater.Run();
+    db_slice.Del(op_args.db_cntx, it->it);
+    if (op_args.shard->journal()) {
+      RecordJournal(op_args, "DEL"sv, {key});
+    }
+  }
+}
+
+bool ContainsWithCleanup(HSetCleanupCtx ctx, const std::string_view field) {
+  const bool found = ctx.sm->Contains(field);
+  ctx.DelKeyIfEmpty();
```

Comment:
This is a blocking function as RecordJournal can preempt
1. Lets rename this function to have suffix Blocking
2. If the key was deleted we should not access sm anymore

Reply:
if I am not wrong with number 2, we need to find a way to enforce that callers will not access sm after calling this functions

Reply:
Now we set `sm` to nullptr if the key is deleted.

Reply:
Another approach could be to consume the input pointer (by always setting to null) and return the string map pointer wrapped in a type such as variant or optional to signify that the string map is no longer valid, but nothing guarantees that the caller will not attempt to use the returned pointer without a check.